texar/modules/connectors/connectors.py

```diff
@@ -470,30 +502,47 @@ class ReparameterizedStochasticConnector(ConnectorBase):
             a ``tuple`` of ``torch.Size``.
             For example, to transform inputs to have decoder state size, set
             :python:`output_size=decoder.state_size`.
+        distribution: A instance of subclass of
+            :torch:`distributions.distribution.Distribution`,
+            Can be a distribution class instance.
+        num_samples (optional): An ``int`` or ``int`` ``Tensor``.
+            Number of samples to generate. If not given,
+            generate a single sample. Note that if batch size has
+            already been included in :attr:`distribution`'s dimensionality,
+            :attr:`num_samples` should be left as ``None``.
         hparams (dict, optional): Hyperparameters. Missing
             hyperparamerter will be set to default values. See
             :meth:`default_hparams` for the hyperparameter sturcture and
             default values.
-        distribution: A instance of subclass of
-            :torch:`distributions.distribution.Distribution`,
-            Can be a class, its name or module path, or a class instance.
-        distribution_kwargs (dict, optional): Keyword arguments for the
-            distribution constructor. Ignored if :attr:`distribution` is a
-            class instance.
     """
 
     def __init__(self,
                  output_size: OutputSize,
-                 hparams: HParamsType = None,
-                 distribution: Union[Type[T], T, str] = 'MultivariateNormal',
-                 distribution_kwargs: Optional[Dict[str, Any]] = None,):
+                 distribution: T,
+                 num_samples: Optional[Union[int, torch.Tensor]] = None,
+                 hparams: Optional[HParamsType] = None):
         ConnectorBase.__init__(self, output_size, hparams)
-        if distribution_kwargs is None:
-            distribution_kwargs = {}
-        self._dstr = utils.check_or_get_instance(
-            distribution, distribution_kwargs,
-            ["torch.distributions", "torch.distributions.multivariate_normal",
-             "texar.custom"])
+
+        self._dstr = distribution
+        for dstr_attr in self._dstr.arg_constraints.keys():  # type: ignore
+            tensor = getattr(self._dstr, dstr_attr)
+            self.register_buffer(dstr_attr, nn.Parameter(tensor))
+            setattr(self._dstr, dstr_attr, getattr(self, dstr_attr))
```

Comment:
Three mistakes here:
1. A parameter should not be registered as a buffer. Use `register_parameter` instead.
2. `register_*` stores the buffer as an attribute internally, so there's no need to call `setattr` here.
3. Implementations of the PyTorch distributions are more convoluted than I thought. To name a few:
   - Actual stored arguments might be different to what's stored inside `arg_constraints`. For instance, `Chi2` inherits from `Gamma` and has argument `df`, however the actual stored attributes are `concentration` and `rate`, which are arguments of the constructor of `Gamma`. 
   - `MultivariateNormal` has four arguments `loc`, `scale_tril`, `covariance_matrix`, and `precision_matrix`, but exactly one among the latter three should be provided. Internally everything is converted to a `scale_tril` matrix, but if the user chooses to parameterize `covariance_matrix` instead of `scale_tril`, we should not make `scale_tril` a parameter.

Sorry for not noticing these details before. Given such absurd implementation of PyTorch distributions, I think the best way to implement this Connector module is to manually support a couple of most commonly used distributions.

To still support arbitrary distributions, I think what we can do is to ask the user to provide the distribution type, arguments to fill in, and initial tensor values for those arguments. Then, we construct the distribution every time `forward` is called.